pkg/reconciler/apiserversource/apiserversource.go

```diff
@@ -0,0 +1,276 @@
+	apiserversourceInformer.Informer().AddEventHandler(reconciler.Handler(impl.Enqueue))
+
+	deploymentInformer.Informer().AddEventHandler(cache.FilteringResourceEventHandler{
+		FilterFunc: controller.Filter(v1alpha1.SchemeGroupVersion.WithKind("ApiServerSource")),
+		Handler:    reconciler.Handler(impl.EnqueueControllerOf),
+	})
+
+	return impl
+}
+
+// Reconcile compares the actual state with the desired, and attempts to
+// converge the two. It then updates the Status block of the ApiServerSource
+// resource with the current status of the resource.
+func (r *Reconciler) Reconcile(ctx context.Context, key string) error {
+	// Convert the namespace/name string into a distinct namespace and name
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		r.Logger.Errorf("invalid resource key: %s", key)
+		return nil
+	}
+
+	// Get the ApiServerSource resource with this namespace/name
+	original, err := r.apiserversourceLister.ApiServerSources(namespace).Get(name)
+	if apierrors.IsNotFound(err) {
+		// The resource may no longer exist, in which case we stop processing.
+		logging.FromContext(ctx).Error("ApiServerSource key in work queue no longer exists", zap.Any("key", key))
+		return nil
+	} else if err != nil {
+		return err
+	}
+
+	// Don't modify the informers copy
+	apiserversource := original.DeepCopy()
+
+	// Reconcile this copy of the ApiServerSource and then write back any status
+	// updates regardless of whether the reconcile error out.
+	err = r.reconcile(ctx, apiserversource)
+	if err != nil {
+		logging.FromContext(ctx).Warn("Error reconciling ApiServerSource", zap.Error(err))
+	} else {
+		logging.FromContext(ctx).Debug("ApiServerSource reconciled")
+		r.Recorder.Eventf(apiserversource, corev1.EventTypeNormal, apiserversourceReconciled, `ApiServerSource reconciled: "%s/%s"`, apiserversource.Namespace, apiserversource.Name)
+	}
+
+	if _, updateStatusErr := r.updateStatus(ctx, apiserversource.DeepCopy()); updateStatusErr != nil {
+		logging.FromContext(ctx).Warn("Failed to update the ApiServerSource", zap.Error(err))
+		r.Recorder.Eventf(apiserversource, corev1.EventTypeWarning, apiserversourceUpdateStatusFailed, "Failed to update ApiServerSource's status: %v", err)
+		return updateStatusErr
+	}
+
+	// Requeue if the resource is not ready:
+	return err
+}
+
+func (r *Reconciler) reconcile(ctx context.Context, source *v1alpha1.ApiServerSource) error {
+	source.Status.InitializeConditions()
+
+	sinkURI, err := duck.GetSinkURI(ctx, r.DynamicClientSet, source.Spec.Sink, source.Namespace)
+	if err != nil {
+		source.Status.MarkNoSink("NotFound", "")
+		return err
+	}
+	source.Status.MarkSink(sinkURI)
+
+	_, err = r.createReceiveAdapter(ctx, source, sinkURI)
+	if err != nil {
+		r.Logger.Error("Unable to create the receive adapter", zap.Error(err))
+		return err
+	}
+
+	// Update source status
+	source.Status.MarkDeployed()
+	return nil
+}
+
+func (r *Reconciler) getReceiveAdapterImage() string {
```

Comment:
I'd prefer that we grab this in the main package and fail there if it's not around. Also I'd prefer we pass it into this method directly instead of panicing here.

Reply:
guess this is a general issue applying to other sources, like `cronjob`

Reply:
Yeah, seems bad to me :) We should fix them all.